File: TestParser/TestObj.fs
module TestParser.TestObj

open NUnit.Framework
open CbStyles.Cbon.Parser
open CbStyles.Cbon.Parser.Parser
open CbStyles.Parser
open System.Collections.Generic

[<SetUp>]
let Setup () =
    ()

[<Test>]
let TestObj1 () =
    let code = Reader.reader "{}"
    let r = obj_loop code
    Assert.True(r.IsSome)
    let struct(s, r) = r.Value
    printf "(%s, %s)" (s.ToString()) (CbAst.ObjToString r)
    Assert.AreEqual(Span<Code>([||]), s)
    let e = new Dictionary<string, CbAst>()
    CollectionAssert.AreEqual(e, r)

[<Test>]
let TestObj2 () =
    let code = Reader.reader "{ a 1 }"
    let r = obj_loop code
    Assert.True(r.IsSome)
    let struct(s, r) = r.Value
    printf "(%s, %s)" (s.ToString()) (CbAst.ObjToString r)
    Assert.AreEqual(Span<Code>([||]), s)

    let e = new Dictionary<string, CbAst>()
    e.Add("a", CbAst.Num (ANum "1"))
    CollectionAssert.AreEqual(e, r)

[<Test>]
let TestObj3 () =
    let code = Reader.reader "{ a 1, b true }"
    let r = obj_loop code
    Assert.True(r.IsSome)
    let struct(s, r) = r.Value
    printf "(%s, %s)" (s.ToString()) (CbAst.ObjToString r)
    Assert.AreEqual(Span<Code>([||]), s)

    let e = new Dictionary<string, CbAst>()
    e.Add("a", CbAst.Num (ANum "1"))
    e.Add("b", CbAst.Bool true)
    CollectionAssert.AreEqual(e, r)

[<Test>]
let TestObj4 () =
    let code = Reader.reader "{ a false b null }"
    let r = obj_loop code
    Assert.True(r.IsSome)
    let struct(s, r) = r.Value
    printf "(%s, %s)" (s.ToString()) (CbAst.ObjToString r)
    Assert.AreEqual(Span<Code>([||]), s)

    let e = new Dictionary<string, CbAst>()
    e.Add("a", CbAst.Bool false)
    e.Add("b", CbAst.Null)
    CollectionAssert.AreEqual(e, r)

[<Test>]
let TestObj5 () =
    let code = Reader.reader "{ \"a\": \"123\", 'b' = '321' ; ; c asd }"
    let r = obj_loop code
    Assert.True(r.IsSome)
    let struct(s, r) = r.Value
    printf "(%s, %s)" (s.ToString()) (CbAst.ObjToString r)
    Assert.AreEqual(Span<Code>([||]), s)

    let e = new Dictionary<string, CbAst>()
    e.Add("a", CbAst.Str "123")
    e.Add("b", CbAst.Str "321")
    e.Add("c", CbAst.Str "asd")
    CollectionAssert.AreEqual(e, r)

[<Test>]
let TestObj6 () =
    let code = Reader.reader "{ a }"
    let f () = obj_loop code |> ignore
    let e = Assert.Throws<ParserError> (new TestDelegate(f))
    printf "%s" (e.Message)

[<Test>]
let TestObj7 () =
    let code = Reader.reader "{ : }"
    let f () = obj_loop code |> ignore
    let e = Assert.Throws<ParserError> (new TestDelegate(f))
    printf "%s" (e.Message)

[<Test>]
let TestObj8 () =
    let code = Reader.reader "{"
    let r = obj_loop code
    Assert.True(r.IsSome)
    let struct(s, r) = r.Value
    printf "(%s, %s)" (s.ToString()) (CbAst.ObjToString r)
    Assert.AreEqual(Span<Code>([||]), s)
    let e = new Dictionary<string, CbAst>()
    CollectionAssert.AreEqual(e, r)

File: TestParser/TestObj.cs
using CbStyles.Cbon;
using CbStyles.Cbon.Parser;
using NUnit.Framework;
using System.Collections.Generic;


namespace TestParser
{
    public class TestObj
    {

        [SetUp]
        public void Setup() { }

        [Test]
        public void TestObj1()
        {
            var code = "{}";
            Parser.RunInReader(code, code => {
                var r = Parser.ObjLoop(code).Value.Item2;
                Assert.AreEqual(new Dictionary<string, CbVal>(), r);
            });
        }

        [Test]
        public void TestObj2()
        {
            var code = "{ a 1 }";
            Parser.RunInReader(code, code => {
                var r = Parser.ObjLoop(code).Value.Item2;
                Assert.AreEqual(new Dictionary<string, CbVal>() { { "a", CbVal.NewNum("1") } }, r);
            });
        }

        [Test]
        public void TestObj3()
        {
            var code = "{ a 1, b true }";
            Parser.RunInReader(code, code => {
                var r = Parser.ObjLoop(code).Value.Item2;
                Assert.AreEqual(new Dictionary<string, CbVal>() { { "a", CbVal.NewNum("1") }, { "b", CbVal.NewBool(true) } }, r);
            });
        }

        [Test]
        public void TestObj4()
        {
            var code = "{ a false b null }";
            Parser.RunInReader(code, code => {
                var r = Parser.ObjLoop(code).Value.Item2;
                Assert.AreEqual(new Dictionary<string, CbVal>() { { "a", CbVal.NewBool(false) }, { "b", CbVal.NewNull() } }, r);
            });
        }

        [Test]
        public void TestObj5()
        {
            var code = "{ \"a\": \"123\", 'b' = '321' ; ; c asd }";
            Parser.RunInReader(code, code => {
                var r = Parser.ObjLoop(code).Value.Item2;
                Assert.AreEqual(new Dictionary<string, CbVal>() { { "a", CbVal.NewStr("123") }, { "b", CbVal.NewStr("321") }, { "c", CbVal.NewStr("asd") } }, r);
            });
        }

        [Test]
        public void TestObj6()
        {
            var code = "{ a }";
            var e = Assert.Throws<ParserException>(() => Parser.RunInReader(code, code => {
                Parser.ObjLoop(code);
            }));
            Assert.AreEqual("Expected object value but not found \t at 1:5", e.Message);
        }

        [Test]
        public void TestObj7()
        {
            var code = "{ : }";
            var e = Assert.Throws<ParserException>(() => Parser.RunInReader(code, code => {
                Parser.ObjLoop(code);
            }));
            Assert.AreEqual("Expected object tag but not found \t at 1:2", e.Message);
        }

        [Test]
        public void TestObj8()
        {
            var code = "{";
            var e = Assert.Throws<ParserException>(() => Parser.RunInReader(code, code => {
                Parser.ObjLoop(code);
            }));
            Assert.AreEqual("Unexpected EOF \t at 1:1", e.Message);
        }
    }
}
